core/libp2p/crypto/crypto_provider.hpp

```diff
@@ -112,45 +112,47 @@ namespace libp2p::crypto {
      */
     virtual std::vector<common::StretchedKey> keyStretcher(
         common::CipherType cipher_type, common::HashType hash_type,
-        const kagome::common::Buffer &secret) const = 0;
+        const Buffer &secret) const = 0;
 
     /**
      * Convert the public key into Protobuf representation
      * @param key - public key to be mashalled
      * @return bytes of Protobuf object
      */
-    virtual kagome::common::Buffer marshal(const PublicKey &key) const = 0;
+    virtual Buffer marshal(const PublicKey &key) const = 0;
 
     /**
      * Convert the private key into Protobuf representation
      * @param key - public key to be mashalled
      * @return bytes of Protobuf object
      */
-    virtual kagome::common::Buffer marshal(const PrivateKey &key) const = 0;
+    virtual Buffer marshal(const PrivateKey &key) const = 0;
 
     /**
      * Convert Protobuf representation of public key into the object
      * @param key_bytes - bytes of the public key
      * @return public key in case of success, none otherwise
      */
-    virtual std::optional<PublicKey> unmarshalPublicKey(
-        const kagome::common::Buffer &key_bytes) const = 0;
+    virtual std::shared_ptr<PublicKey> unmarshalPublicKey(
+        const Buffer &key_bytes) const = 0;
 
     /**
      * Convert Protobuf representation of private key into the object
      * @param key_bytes - bytes of the private key
      * @return private key in case of success, none otherwise
      */
-    virtual std::optional<PrivateKey> unmarshalPrivateKey(
-        const kagome::common::Buffer &key_bytes) const = 0;
+    virtual std::shared_ptr<PrivateKey> unmarshalPrivateKey(
```

Comment:
what was wrong with returning optional?

Reply:
optional cannot be instantiated from abstract classes